logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-        resolveLabels.removeIf(label -> traversalEng.graph().schema().getType(label).isAbstract());
-        variable.addResolvedTypes(resolveLabels);
+    private void findVariablesWithNoHints(Conjunction conjunction) {
+        long numOfTypes = traversalEng.graph().schema().stats().thingTypeCount();
+        long numOfConcreteTypes = traversalEng.graph().schema().stats().concreteThingTypeCount();
+        iterate(conjunction.variables()).filter(variable -> (
+                variable.isType() && variable.resolvedTypes().size() == numOfTypes ||
+                        variable.isThing() && variable.resolvedTypes().size() == numOfConcreteTypes
+        )).forEachRemaining(Variable::clearResolvedTypes);
     }
 
-    private Map<Reference, Set<Label>> retrieveResolveTypes(Set<Variable> resolveVars) {
-        Conjunction resolveVariableConjunction = new Conjunction(resolveVars, Collections.emptySet());
-        resolveVariableConjunction = resolveLabels(resolveVariableConjunction);
-        return logicCache.resolver().get(resolveVariableConjunction, conjunction -> {
+    private Map<Reference, Set<Label>> runTraversalEngine(TraversalConstructor traversalConstructor) {
+        return logicCache.resolver().get(traversalConstructor.resolverTraversal(), traversal -> {
             Map<Reference, Set<Label>> mapping = new HashMap<>();
-            traversalEng.iterator(conjunction.traversal()).forEachRemaining(
+            traversalEng.iterator(traversal).forEachRemaining(
                     result -> result.forEach((ref, vertex) -> {
                         mapping.putIfAbsent(ref, new HashSet<>());
-                        mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
+                        assert vertex.isType();
+                        if (!(vertex.asType().isAbstract() && traversalConstructor.getVariable(ref).isThing()))
+                            mapping.get(ref).add(Label.of(vertex.asType().label(), vertex.asType().scope()));
                     })
             );
             return mapping;
         });
     }
 
-    private Type getType(Label label) {
-        if (label.scope().isPresent()) {
-            assert conceptMgr.getRelationType(label.scope().get()) != null;
-            return conceptMgr.getRelationType(label.scope().get()).getRelates(label.name());
-        } else {
-            return conceptMgr.getThingType(label.name());
-        }
-    }
+    private static class TraversalConstructor {
 
-    private static class ConstraintMapper {
-        private final Resolvers resolvers;
-        private final HashMap<TypeVariable, Set<TypeVariable>> neighbours;
-        private final TypeVariable rootAttributeType;
-        private final TypeVariable rootRelationType;
-        private final TypeVariable rootRoleType;
         private final Conjunction conjunction;
+        private final Map<Reference, Variable> variableRegister;
```

Comment:
As we meet variables in the conjunction we register the their references. This is so we can refer to the correct variables from references after the `TraversalEngine` has been run.